docs: split out files and folders and the last few attributes (#15004)
conan-io/conan-center-index: docs/adding_packages/conanfile_attributes.md

<!-- toc -->
## Contents

  * [Attributes](#attributes)
    * [Name](#name)
    * [Version](#version)
    * [ConanCenter specific releases format](#conancenter-specific-releases-format)
    * [License Attribute](#license)
  * [Settings](#settings)
  * [Options](#options)
    * [Recommended Names](#recommended-names)
    * [Predefined Options and Known Defaults](#predefined-options-and-known-defaults)
    * [Options to Avoid](#options-to-avoid)<!-- endToc -->

## Attributes

> There's are a [key features](https://docs.conan.io/en/latest/reference/conanfile/attributes.html) which allow the Conan client to understand,

Reviewer: ```suggestion
There are a [key features](https://docs.conan.io/en/latest/reference/conanfile/attributes.html) which allow the Conan client to understand,
```

Author: ```suggestion
These are a [key feature](https://docs.conan.io/en/latest/reference/conanfile/attributes.html) which allow the Conan client to understand,
```

identify, and expose recipes and which project they expose.

> In ConanCenter, there are a few conventions that need to be respected to ensure recipes can be discovered there `conan search` command

Reviewer 2: ```suggestion
In ConanCenter, there are a few conventions that need to be respected to ensure recipes can be discovered with the `conan search` command
```

> of through the web UI. Many of which are enforces with the [hooks](../error_knowledge_base.md).

Reviewer 2: ```suggestion
through the web UI. Many of these are enforced with the [hooks](../error_knowledge_base.md).
```

### Name

Same as the _recipe folder_ and always lowercase.

Please see the FAQs for:

* [name collisions](../faqs.md#what-is-the-policy-on-recipe-name-collisions)
* [space and symbols](../faqs.md#should-reference-names-use---or-_)

### Version

> ConanCenter is geared towards quickly added new release, the build service always pass the version it is currently building to the recipe.

Reviewer 2: ```suggestion
ConanCenter is geared towards quickly adding a new release, the build service always passes the version it is currently building to the recipe.
```

> The `version` attribute MUST NOT be added to any recipe - with except to "system packages".

Reviewer: ```suggestion
The `version` attribute MUST NOT be added to any recipe - with exception to "system packages".
```

> The `version` attribute MUST NOT be added to any recipe - with exception to "system packages".

Reviewer 2: ```suggestion
The `version` attribute MUST NOT be added to any recipe - with the exception of "system packages".
```

#### ConanCenter specific releases format

> The notation shown below is used for publishing packages which do not match the original library's official releases. This format which includes the "datestamp" corresponding to the date of a commit: `cci.<YEAR MONTH DAY>`.

Reviewer 2: ```suggestion
The notation shown below is used for publishing packages that do not match the original library's official releases. This format includes the "datestamp" corresponding to the date of a commit: `cci.<YEAR MONTH DAY>`.
```

In order to create reproducible builds, we also "commit-lock" to the latest commit on that day. Otherwise, users would get inconsistent results over time when rebuilding the package. An example of this is the [RapidJSON](https://github.com/Tencent/rapidjson) library, where its package reference is `rapidjson/cci.20200410` and its sources are locked the latest commit on that date in [config.yml](https://github.com/conan-io/conan-center-index/blob/master/recipes/rapidjson/config.yml#L4). The prefix `cci.` is mandatory to distinguish as a virtual version provided by CCI. If you are interested to know about the origin, please, read [here](https://github.com/conan-io/conan-center-index/pull/1464).

### License Attribute

The mandatory license attribute of each recipe **should** be a [SPDX license](https://spdx.org/licenses/) [short Identifiers](https://spdx.dev/ids/) when applicable.

Where the SPDX guidelines do not apply, packages should do the following:

> * When no license is provided or it's under the ["public domain"](https://fairuse.stanford.edu/overview/public-domain/welcome/) - these are not a license by itself. Thus, we have [equivalent licenses](https://en.wikipedia.org/wiki/Public-domain-equivalent_license) that should be used instead. If a project fall under these criteria it should be identified as the [Unlicense](https://spdx.org/licenses/Unlicense) license.

Reviewer: ```suggestion
* When no license is provided or it's under the ["public domain"](https://fairuse.stanford.edu/overview/public-domain/welcome/) - these are not a license by itself. Thus, we have [equivalent licenses](https://en.wikipedia.org/wiki/Public-domain-equivalent_license) that should be used instead. If a project falls under these criteria it should be identified as the [Unlicense](https://spdx.org/licenses/Unlicense) license.
```

> * When no license is provided or it's under the ["public domain"](https://fairuse.stanford.edu/overview/public-domain/welcome/) - these are not a license by itself. Thus, we have [equivalent licenses](https://en.wikipedia.org/wiki/Public-domain-equivalent_license) that should be used instead. If a project falls under these criteria it should be identified as the [Unlicense](https://spdx.org/licenses/Unlicense) license.

Reviewer 2: ```suggestion
* When no license is provided or it's under the ["public domain"](https://fairuse.stanford.edu/overview/public-domain/welcome/) - this is not a license by itself. Thus, we have [equivalent licenses](https://en.wikipedia.org/wiki/Public-domain-equivalent_license) that should be used instead. If a project falls under these criteria it should be identified as the [Unlicense](https://spdx.org/licenses/Unlicense) license.
```

* When a custom (e.g. project specific) license is given, the value should be set to `LicenseRef-` as a prefix, followed by the name of the file which contains the custom license. See [this example](https://github.com/conan-io/conan-center-index/blob/e604534bbe0ef56bdb1f8513b83404eff02aebc8/recipes/fft/all/conanfile.py#L8). For more details, [read this conversation](https://github.com/conan-io/conan-center-index/pull/4928/files#r596216206).

## Order of methods and attributes

Prefer the following order of documented methods in python code (`conanfile.py`, `test_package/conanfile.py`):

For `conan create` the order is listed [here](https://docs.conan.io/en/latest/reference/commands/creator/create.html#methods-execution-order)
test packages recipes should append the following methods:

* deploy
* test

> the order above resembles the execution order of methods on CI. therefore, for instance, `build` is always executed before `package` method, so `build` should appear before the

Reviewer 2: ```suggestion
The order above resembles the execution order of methods on CI. Therefore, for instance, `build` is always executed before the `package` method, so `build` should appear before the
```
